lib/ProgramCall.js

```diff
@@ -31,59 +33,108 @@ class ProgramCall {
   }
 
   /**
-    * @description adds a parameter to the program XML
-    * @param {string | array} data
-    * @param {string} [type]
-    * @param {object} options
-    * @param {*} inDs
-    */
-  addParam(data, type, options, inDs = null) {
-    let opt;
-    // DS element has no 'type', so if the second param 'type' is an Object, then it is the options.
-    if (Array.isArray(data)) {
-      opt = type;
-    } else {
-      opt = options;
-      if (type === undefined) {
-        throw new Error('Specifying the parameter type is required.');
-      }
+   * Internal function to create <data> nodes
+   * There is an open propsal to add private methods to JS.
+   * https://github.com/tc39/proposal-private-methods
+   * We should update to use private methods in the future.
+   * @param {string} name
+   * @param {string} type
+   * @param {string} value
+   * @param {object} options
+  */
+  __addDataNode__(name, type, value, options) {
+    const dataName = name ? ` name='${name}'` : '';
+    this.xml += `<data${dataName} type='${type}'`;
+    // append <data> options
+    if (typeof options === 'object') {
+      if (options.varying) { this.xml += ` varying='${options.varying.varying}'`; }
+      if (options.enddo) { this.xml += ` enddo='${options.enddo}'`; }
+      if (options.setlen) { this.xml += ` setlen='${options.setlen}'`; }
+      if (options.offset) { this.xml += ` offset='${options.offset}'`; }
+      if (options.hex) { this.xml += ` hex='${options.hex}'`; }
+      if (options.trim) { this.xml += ` trim='${options.trim}'`; }
+      if (options.next) { this.xml += ` next='${options.next}'`; }
     }
-    if (!inDs) { // In recursive mode, if it is an element in DS, then no <parm> or </parm> needed.
-      this.xml += '<parm';
-      if (opt && typeof opt === 'object') { // append <param> options
-        if (opt.io) { this.xml += ` io='${opt.io}'`; }
-        if (opt.by) { this.xml += ` by='${opt.by}'`; }
+    this.xml += `>${value}</data>`;
+  }
+
+  /**
+   * Internal function to create <ds> nodes
+   * There is an open propsal to add private methods to JS.
+   * https://github.com/tc39/proposal-private-methods
+   * We should update to use private methods in the future.
+   * @param {array} ds
+   * @param {object} options
+   */
+  __addDsNodes__(ds, options = {}) {
```

Comment:
I don't see why this function is needed. It should be rolled in to `__addData__`.